src/hooks/useGame.tsx

```diff
@@ -1,134 +1,172 @@
+
+        // 점수 연산
+        const newScore = getScore(newCells);
+        const newHighScore = getHighScore(newScore, highScore);
+
+        // 다음 턴 진행 가능한지 확인
+        if (is128Exist(newCells)) {
+          setGameOver(GameOverStatus.Success);
+          saveLocalStorage({ gameOver: GameOverStatus.Success });
+        } else if (!checkCanMove(newCells)) {
+          setGameOver(GameOverStatus.Fail);
+          saveLocalStorage({ gameOver: GameOverStatus.Fail });
+        }
+
+        // 데이터 저장
+        saveLocalStorage({
+          cells: newCells,
+          history: [...history, structuredClone(newCells)],
+          score: newScore,
+          highScore: newHighScore,
+        });
+
+        setCells(newCells);
+        saveCellsHistory(newCells);
+        setScore(newScore);
+        setHighScore(newHighScore);
+      }
     },
-    [moveLeft, rotateMapCounterClockwise],
+    [cells, history, highScore, saveCellsHistory],
   );
 
-  const is128Exist = useCallback((newCells: (number | null)[][]) => {
-    return newCells.some((row) => row.includes(128));
-  }, []);
+  const checkUndo = () => {
+    const undoResult = undo(history);
 
-  const checkNextTurn = useCallback(() => {
-    const emptyCells = getEmptyCellsIndex();
+    if (undoResult.currentCell !== undefined) {
+      setCells(undoResult.currentCell);
+      // 점수 변경
+      const newScore = getScore(undoResult.currentCell);
+      setScore(newScore);
+
+      // 변경 내용 로컬스토리지에 반영
+      saveLocalStorage({
+        cells: undoResult.currentCell,
+        history: undoResult.history,
+      });
+    }
+    setHistory(undoResult.history);
+  };
 
-    if (emptyCells.length === 0) return;
+  const checkReload = useCallback(() => {
+    const data = getLocalData();
 
-    // 새로운 셀 생성
-    const randomIndex = Math.floor(Math.random() * emptyCells.length);
-    // 이 조건 깔끔하게 처리할 수 있는 방법 확인
-    if (emptyCells[randomIndex] !== undefined) {
-      const [rowIndex, colIndex] = emptyCells[randomIndex];
+    const emptyCells = Array<null>(4)
+      .fill(null)
+      .map(() => Array<null>(4).fill(null));
+    const initCells = addTwoRandomCells(emptyCells);
 
-      const newCells = cells.map((row) => [...row]);
+    // 로컬스토리지에 데이터가 있으면 해당 데이터로 불러오기
+    if (data.cells !== undefined) {
+      setCells(data.cells);
+    } else {
+      data.cells = initCells;
+      setCells(initCells);
+    }
+    if (data.history !== undefined) {
+      setHistory(data.history);
+    } else {
+      data.history = [];
+      setHistory([]);
+    }
+    if (data.score !== undefined) {
+      setScore(data.score);
+    } else {
+      data.score = 4;
+      setScore(4);
+    }
+    if (data.highScore !== undefined) {
+      setHighScore(data.highScore);
+    } else {
+      data.highScore = 4;
+      setHighScore(4);
+    }
+    if (data.gameOver !== undefined) {
+      setGameOver(data.gameOver);
+    } else {
+      data.gameOver = GameOverStatus.None;
+      setGameOver(GameOverStatus.None);
+    }
```

Comment:
`saveLocalStorage` 를 이동시에도 체크하고 Init에도 체크하고 Reload에도 체크하고 있는데, 좀더 깔끔하게 useEffect 하나 새로 만들어서

```ts
useEffect(() => saveLocalStorage({ highScore, cells, history, score, gameOver });
```

해도 괜찮을 것 같습니다 👀 

그럼 이 로직들 다 정리할 수 있을 것 같아요 (다른 곳들도 마찬가지)

```suggestion
    setCells(data.cells ?? initCells);
    setHistory(data.history ?? []);
    setScore(data.score ?? 4);
    setHighScore(data.highScore ?? 4);
    setGameOver(data.gameOver ?? GameOverStatus.None);
```